lib/test/mock_loader.js

```diff
@@ -0,0 +1,228 @@
+'use strict';
+const {
+  SafeMap,
+  globalThis,
+} = primordials;
+const {
+  Atomics: {
+    notify: AtomicsNotify,
+    store: AtomicsStore,
+  },
+} = globalThis;
+const {
+  ensureNodeScheme,
+  kBadExportsMessage,
+  kMockSearchParam,
+  kMockSuccess,
+  kMockExists,
+  kMockUnknownMessage,
+} = require('internal/test_runner/mock/mock');
+const {
+  pathToFileURL,
+  URL,
+} = require('internal/url');
+const { normalizeReferrerURL } = require('internal/modules/helpers');
+let debug = require('internal/util/debuglog').debuglog('test_runner', (fn) => {
+  debug = fn;
+});
+const { createRequire, isBuiltin } = require('module');
+
+// TODO(cjihrig): This file should not be exposed publicly, but register() does
+// not handle internal loaders. Before marking this API as stable, one of the
+// following issues needs to be implemented:
+// https://github.com/nodejs/node/issues/49473
+// or https://github.com/nodejs/node/issues/52219
+
+// TODO(cjihrig): The mocks need to be thread aware because the exports are
+// evaluated on the thread that creates the mock. Before marking this API as
+// stable, one of the following issues needs to be implemented:
+// https://github.com/nodejs/node/issues/49472
+// or https://github.com/nodejs/node/issues/52219
+
+// TODO(cjihrig): Network imports should be supported. There are two current
+// hurdles:
+// - The module format returned by the load() hook is not known. This could be
+//   implemented as an option, or default to 'module' for network imports.
+// - The generated mock module imports 'node:test', which is not allowed by
+//   checkIfDisallowedImport() in the ESM code.
+
+const mocks = new SafeMap();
+
+async function initialize(data) {
+  data?.port.on('message', ({ type, payload }) => {
+    debug('mock loader received message type "%s" with payload %o', type, payload);
+
+    if (type === 'node:test:register') {
+      const { baseURL } = payload;
+      const mock = mocks.get(baseURL);
+
+      if (mock?.active) {
+        debug('already mocking "%s"', baseURL);
+        sendAck(payload.ack, kMockExists);
+        return;
+      }
+
+      const localVersion = mock?.localVersion ?? 0;
+
+      debug('new mock version %d for "%s"', localVersion, baseURL);
+      mocks.set(baseURL, {
+        __proto__: null,
```

Comment:
is there a real risk of prototype pollution within this hook?

Reply:
Probably not.